File: store.py
"""Commit a full CAPE attempt record to GitHub as the audit trail."""

from __future__ import annotations

import base64
import json
import os
import re
from datetime import datetime, timezone
from typing import Any

import requests

from llm import github_token
from report import render_report

DEFAULT_REPO = "KV-LD/gama-cape-prototype"


def _slug(text: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9]+", "_", text or "candidate").strip("_")
    return cleaned[:80] or "candidate"


def _put_file(
    *,
    repo: str,
    path: str,
    content: str,
    message: str,
    token: str,
) -> dict[str, Any]:
    encoded = base64.b64encode(content.encode("utf-8")).decode("ascii")
    url = f"https://api.github.com/repos/{repo}/contents/{path}"
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "gama-cape-prototype",
    }
    payload = {
        "message": message,
        "content": encoded,
        "branch": os.getenv("GITHUB_ATTEMPTS_BRANCH", "main"),
    }
    resp = requests.put(url, headers=headers, json=payload, timeout=60)
    if resp.status_code not in (200, 201):
        raise RuntimeError(
            f"GitHub would not save {path} (HTTP {resp.status_code}). "
            f"GITHUB_TOKEN found but cannot write to {repo} — check the token's repo permissions. "
            f"Details: {resp.text[:400]}"
        )
    return resp.json()


def _content_html_url(data: dict[str, Any], repo: str, path: str) -> str:
    html_url = (data.get("content") or {}).get("html_url") or data.get("html_url")
    if html_url:
        return html_url
    branch = os.getenv("GITHUB_ATTEMPTS_BRANCH", "main")
    return f"https://github.com/{repo}/blob/{branch}/{path}"


def commit_attempt(record: dict[str, Any], repo: str = DEFAULT_REPO) -> str:
    repo = os.getenv("GITHUB_REPO", repo)
    token = github_token()
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    iso_time = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    candidate = _slug(str(record.get("candidate_name") or "candidate"))
    module_id = record.get("module_id") or record.get("module", {}).get("id") or "x"
    folder = f"attempts/{candidate}_{module_id}_{timestamp}"
    json_path = f"{folder}/record.json"
    html_path = f"{folder}/report.html"

    record = dict(record)
    record.setdefault("recorded_at", iso_time)

    verdict = record.get("verdict") or (record.get("score_sheet") or {}).get("verdict") or "UNKNOWN"
    module_name = record.get("module_name") or ""
    message = f"CAPE attempt: {record.get('candidate_name')} - Module {module_id} ({module_name}) - {verdict}"

    body_json = json.dumps(record, indent=2, ensure_ascii=False)
    json_resp = _put_file(repo=repo, path=json_path, content=body_json, message=message, token=token)
    github_url = _content_html_url(json_resp, repo, json_path)

    record["github_url"] = github_url
    html_doc = render_report(record)
    html_resp = _put_file(
        repo=repo,
        path=html_path,
        content=html_doc,
        message=f"{message} (HTML report)",
        token=token,
    )
    html_report_url = _content_html_url(html_resp, repo, html_path)

    # Refresh JSON so the stored record includes both URLs.
    record["html_report_url"] = html_report_url
    body_json = json.dumps(record, indent=2, ensure_ascii=False)
    sha = ((json_resp.get("content") or {}).get("sha")) or json_resp.get("sha")
    url = f"https://api.github.com/repos/{repo}/contents/{json_path}"
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "gama-cape-prototype",
    }
    encoded = base64.b64encode(body_json.encode("utf-8")).decode("ascii")
    payload = {
        "message": f"{message} (audit URLs)",
        "content": encoded,
        "branch": os.getenv("GITHUB_ATTEMPTS_BRANCH", "main"),
        "sha": sha,
    }
    resp = requests.put(url, headers=headers, json=payload, timeout=60)
    if resp.status_code not in (200, 201):
        return {"github_url": github_url, "html_report_url": html_report_url}
    return {"github_url": github_url, "html_report_url": html_report_url}

File: lib/store.ts
import { DEFAULT_REPO } from "./constants";
import { githubToken } from "./llm";
import { renderReport } from "./report";
import type { AttemptRecord } from "./types";

function slug(text: string): string {
  const cleaned = (text || "candidate").replace(/[^A-Za-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
  return (cleaned.slice(0, 80) || "candidate");
}

async function putFile(repo: string, path: string, content: string, message: string, sha?: string) {
  const token = githubToken();
  const payload: Record<string, unknown> = {
    message,
    content: Buffer.from(content, "utf8").toString("base64"),
    branch: process.env.GITHUB_ATTEMPTS_BRANCH || "main",
  };
  if (sha) payload.sha = sha;
  const resp = await fetch(`https://api.github.com/repos/${repo}/contents/${path}`, {
    method: "PUT",
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28",
      "User-Agent": "gama-cape-prototype",
    },
    body: JSON.stringify(payload),
  });
  if (!resp.ok) {
    const details = (await resp.text()).slice(0, 400);
    throw new Error(
      `GitHub would not save ${path} (HTTP ${resp.status}). Check GITHUB_TOKEN repo permissions. Details: ${details}`,
    );
  }
  return (await resp.json()) as {
    content?: { html_url?: string; sha?: string };
    html_url?: string;
    sha?: string;
  };
}

function contentUrl(
  data: { content?: { html_url?: string }; html_url?: string },
  repo: string,
  path: string,
): string {
  return (
    data.content?.html_url ||
    data.html_url ||
    `https://github.com/${repo}/blob/${process.env.GITHUB_ATTEMPTS_BRANCH || "main"}/${path}`
  );
}

export async function commitAttempt(record: AttemptRecord): Promise<{
  github_url: string;
  html_report_url: string;
}> {
  const repo = process.env.GITHUB_REPO || DEFAULT_REPO;
  const now = new Date();
  const timestamp = now.toISOString().replace(/[-:]/g, "").replace(/\.\d{3}Z$/, "Z");
  const isoTime = now.toISOString().replace(/\.\d{3}Z$/, "Z");
  const candidate = slug(String(record.candidate_name || "candidate"));
  const moduleId = record.module_id ?? "x";
  const folder = `attempts/${candidate}_${moduleId}_${timestamp}`;
  const jsonPath = `${folder}/record.json`;
  const htmlPath = `${folder}/report.html`;

  record.recorded_at = record.recorded_at || isoTime;
  const verdict = record.verdict || record.score_sheet?.verdict || "UNKNOWN";
  const message = `CAPE attempt: ${record.candidate_name} - Module ${moduleId} (${record.module_name}) - ${verdict}`;

  const jsonResp = await putFile(repo, jsonPath, JSON.stringify(record, null, 2), message);
  const githubUrl = contentUrl(jsonResp, repo, jsonPath);
  record.github_url = githubUrl;

  const htmlDoc = renderReport(record);
  const htmlResp = await putFile(repo, htmlPath, htmlDoc, `${message} (HTML report)`);
  const htmlReportUrl = contentUrl(htmlResp, repo, htmlPath);
  record.html_report_url = htmlReportUrl;

  const sha = jsonResp.content?.sha || jsonResp.sha;
  await putFile(repo, jsonPath, JSON.stringify(record, null, 2), `${message} (audit URLs)`, sha);

  return { github_url: githubUrl, html_report_url: htmlReportUrl };
}
